core/drivers/clk/clk.c

```diff
@@ -0,0 +1,273 @@
+{
+	return refcount_val(&clk->enabled_count) != 0;
+}
+
+static TEE_Result clk_enable_no_lock(struct clk *clk)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct clk *parent = NULL;
+
+	if (refcount_inc(&clk->enabled_count))
+		return TEE_SUCCESS;
+
+	parent = clk_get_parent(clk);
+	if (parent) {
+		res = clk_enable_no_lock(parent);
+		if (res)
+			return res;
+	}
+
+	if (clk->ops->enable) {
+		res = clk->ops->enable(clk);
+		if (res)
+			return res;
+	}
+
+	refcount_set(&clk->enabled_count, 1);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_enable(struct clk *clk)
+{
+	uint32_t exceptions = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	exceptions =  cpu_spin_lock_xsave(&clk_lock);
+	res = clk_enable_no_lock(clk);
+	cpu_spin_unlock_xrestore(&clk_lock, exceptions);
+
+	return res;
+}
+
+static void clk_disable_no_lock(struct clk *clk)
+{
+	struct clk *parent = NULL;
+
+	if (!refcount_dec(&clk->enabled_count))
+		return;
+
+	if (clk->ops->disable)
+		clk->ops->disable(clk);
+
+	parent = clk_get_parent(clk);
+	if (parent)
+		clk_disable_no_lock(parent);
+}
+
+void clk_disable(struct clk *clk)
+{
+	uint32_t exceptions = 0;
+
+	exceptions =  cpu_spin_lock_xsave(&clk_lock);
+	clk_disable_no_lock(clk);
+	cpu_spin_unlock_xrestore(&clk_lock, exceptions);
+}
+
+unsigned long clk_get_rate(struct clk *clk)
+{
+	return clk->rate;
+}
+
+static TEE_Result clk_set_rate_no_lock(struct clk *clk, unsigned long rate)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	unsigned long parent_rate = 0;
+
+	if (clk->parent)
+		parent_rate = clk_get_rate(clk->parent);
+
+	res = clk->ops->set_rate(clk, rate, parent_rate);
+	if (res)
+		return res;
+
+	clk_compute_rate_no_lock(clk);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result clk_set_rate(struct clk *clk, unsigned long rate)
+{
+	uint32_t exceptions = 0;
+	TEE_Result res = TEE_ERROR_GENERIC;
+
+	if (!clk->ops->set_rate)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	exceptions =  cpu_spin_lock_xsave(&clk_lock);
+
+	if (clk->flags & CLK_SET_RATE_GATE && clk_is_enabled_no_lock(clk))
+		res = TEE_ERROR_BAD_STATE;
+	else
+		res = clk_set_rate_no_lock(clk, rate);
+
+	cpu_spin_unlock_xrestore(&clk_lock, exceptions);
+
+	return res;
+}
+
+struct clk *clk_get_parent(struct clk *clk)
+{
+	return clk->parent;
+}
+
+static TEE_Result clk_set_parent_no_lock(struct clk *clk, size_t pidx)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	bool was_enabled = false;
+
+	/* Requested parent is already the one set */
+	if (clk->parent == clk->parents[pidx])
+		return TEE_SUCCESS;
+
+	was_enabled = clk_is_enabled_no_lock(clk);
+	/* Call is needed to decrement refcount on current parent tree */
+	if (was_enabled)
+		clk_disable_no_lock(clk);
+
+	res = clk->ops->set_parent(clk, pidx);
+	if (res)
+		goto out;
+
+	clk->parent = clk->parents[pidx];
+
+	/* The parent changed and the rate might also have changed */
+	clk_compute_rate_no_lock(clk);
+
+out:
+	/* Call is needed to increment refcount on the new parent tree */
+	if (was_enabled)
+		res = clk_enable_no_lock(clk);
```

Comment:
I thins should this restore the original parent clock upon failure.

Reply:
After reviewing this part with some hindsight, I'm thinking the pidx parameter for the parent is not really convenient at all. I will probably replace this with the parent clock struct directly (at least for the public part). For the private interface with clocks, it makes more sense since this is often an id that is used.

Reply:
Fixed